### PR TITLE
[vNext] Fixing List dividers: last cell with children shows an additional divider

### PR DESCRIPTION
**Pull request by sophialee0416**
### Platforms Impacted
- [x] iOS
- [ ] macOS

### Description of changes

The last cell in a nested list showed double dividers which did not align with the intended appearance. This fix gets rid of the extra divider in the List.

This change is temporary as we will probably adapt to the new [List Separators](https://developer.apple.com/documentation/swiftui/menu/listrowseparator(_:edges:)) in the future (iOS 15+).

### Verification

Note: The separators are in red solely for verification purposes

| Before                                       | After                                      |
|----------------------------------------------|--------------------------------------------|
| Simulator Screen Shot - iPhone 11 - 2021-07-21 at 16 27 11 | Simulator Screen Shot - iPhone 11 - 2021-07-21 at 16 26 28 |

### Pull request checklist

This PR has considered:
- [ ] Light and Dark appearances
- [ ] VoiceOver and Keyboard Accessibility
- [ ] Internationalization and Right to Left layouts
- [ ] Different resolutions (1x, 2x, 3x)
- [ ] Size classes and window sizes (iPhone vs iPad, notched devices, multitasking, different window sizes, etc)

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/fluentui-apple/pull/629)

### CODE REVIEW

File: ios/FluentUI.Demo/FluentUI.Demo/Demos/ListDemoController.swift [hunk 1]
```diff
@@ -56,12 +57,13 @@ class ListDemoController: DemoController {
...
+            listCell.hasDivider = true
         }
-        children[0].children = subchildren
-        children[0].isExpanded = true
-        children[1].onTapAction = {
-            self.showAlertForAvatarTapped(name: samplePersonas[3].name)
-        }
+        children[1].children = subchildren
+        children[1].isExpanded = true
+//        children[1].onTapAction = {
```
Comment (harrieshin) [inline ios/FluentUI.Demo/FluentUI.Demo/Demos/ListDemoController.swift:64]: fix?

**Review by rdeassis — CHANGES_REQUESTED** [review]
There's a lot of code that's just commented out. Is this a draft PR you'd like to discuss?
If not, the commented code needs to be removed.

**Review by mischreiber — CHANGES_REQUESTED** [review]

File: ios/FluentUI/Vnext/List/List.swift [hunk 2]
```diff
@@ -70,12 +70,32 @@ public struct MSFListView: View {
...
+        if let children = lastCell.children {
+            return lastDivider(findLastCell(children[children.count - 1]))
+        } else {
+            return lastDivider(lastCell)
+        }
+    }
+
+    private func lastDivider(_ cell: MSFListCellState) -> MSFListCellState {
+        let state = cell
+        state.hasDivider = false
```
Comment (harrieshin) [inline ios/FluentUI/Vnext/List/List.swift:83]: why are we setting the divider state to false here? what is this function for?

File: ios/FluentUI/Vnext/List/List.swift [hunk 3]
```diff
@@ -70,12 +70,32 @@ public struct MSFListView: View {
...
+        return state
+    }
+
     private func updateCellDividers() -> [MSFListSectionState] {
         state.sections.forEach { section in
             section.cells.forEach { cell in
                 cell.hasDivider = section.hasDividers
             }
-            section.cells.last?.hasDivider = false
+            if let last = section.cells.last {
```
Comment (harrieshin) [inline ios/FluentUI/Vnext/List/List.swift:92]: how you are organizing the code can be better here. while you are enumerating (recursively or not) the lambda block should be the one setting the state.hasDivider  to false or true depending on it is a last cell.  how it is currently organized doesn't mix and match how you are traversing the tree and changing its values

File: ios/FluentUI.Demo/FluentUI.Demo/Demos/ListDemoController.swift [hunk 4]
```diff
@@ -43,6 +43,7 @@ class ListDemoController: DemoController {
         listCell.onTapAction = {
             self.showAlertForAvatarTapped(name: samplePersonas[4].name)
         }
+        listCell.hasDivider = true
```
Comment (mischreiber) [inline ios/FluentUI.Demo/FluentUI.Demo/Demos/ListDemoController.swift:46]: Are we changing the default behavior of listCell to not have dividers unless explicitly set to true?

Comment (sophialee0416) [reply ios/FluentUI.Demo/FluentUI.Demo/Demos/ListDemoController.swift:46]: Yep!

Comment (harrieshin) [reply ios/FluentUI.Demo/FluentUI.Demo/Demos/ListDemoController.swift:46]: are you documenting that behavior somewhere?

Comment (sophialee0416) [reply ios/FluentUI.Demo/FluentUI.Demo/Demos/ListDemoController.swift:46]: Looks like it got missed on the description, will add to iteration!

File: ios/FluentUI/Vnext/List/ListCell.swift [hunk 5]
```diff
@@ -27,6 +27,8 @@ import SwiftUI
 ///
 /// `onTapAction`: perform an action on a cell tap.
 ///
+/// `hasDivider`:  displays a separator beneath the cell; default is set to `false`.
```
Comment (harrieshin) [inline ios/FluentUI/Vnext/List/ListCell.swift:30]: nit : period

File: ios/FluentUI/Vnext/List/List.swift [hunk 6]
```diff
@@ -70,12 +70,32 @@ public struct MSFListView: View {
...
+    }
+
     private func updateCellDividers() -> [MSFListSectionState] {
         state.sections.forEach { section in
             section.cells.forEach { cell in
                 cell.hasDivider = section.hasDividers
             }
-            section.cells.last?.hasDivider = false
+            if let lastCell = section.cells.last {
+                let adjacent = findLastCell(lastCell)
```
Comment (harrieshin) [inline ios/FluentUI/Vnext/List/List.swift:96]: nit: don't think adjacent is the right name for this variable?

File: ios/FluentUI/Vnext/List/ListCell.swift (hunk 5 above)
Comment (sophialee0416) [reply ios/FluentUI/Vnext/List/ListCell.swift:30]: Should I remove?

File: ios/FluentUI/Vnext/List/List.swift [hunk 7]
```diff
@@ -70,12 +70,32 @@ public struct MSFListView: View {
...
+    /// Finds the last cell directly adjacent to the end of a list section. This is used to remove the redundant separator that is
+    /// inserted between each cell. Used as a fix until we are able to use the new List Separators in iOS 15.
+    private func findLastCell(_ lastCell: MSFListCellState) -> MSFListCellState {
+        if let children = lastCell.children {
+            if lastCell.isExpanded {
+                guard let lastChild = children.last else {
+                    preconditionFailure("Does not contain any children cells")
+                }
+                return findLastCell(lastChild)
+            } else {
```
Comment (harrieshin) [inline ios/FluentUI/Vnext/List/List.swift:82]: couldn't it be just returning lastCell without 2 addiotnal else statements?

File: ios/FluentUI/Vnext/List/ListCell.swift (hunk 5 above)
Comment (harrieshin) [reply ios/FluentUI/Vnext/List/ListCell.swift:30]: I meant there is a semicolon inbetween your sentences.

**Comment by harrieshin** [comment]
can you also make sure all your pr description and screenshots are updated appropriately based on the template?

File: ios/FluentUI/Vnext/List/List.swift [hunk 8]
```diff
@@ -70,12 +70,32 @@ public struct MSFListView: View {
...
+        }
+    }
+
     private func updateCellDividers() -> [MSFListSectionState] {
         state.sections.forEach { section in
             section.cells.forEach { cell in
                 cell.hasDivider = section.hasDividers
             }
-            section.cells.last?.hasDivider = false
+            if let lastCell = section.cells.last {
```
Comment (harrieshin) [inline ios/FluentUI/Vnext/List/List.swift:95]: also don't we only need to go through this traversing when this section has dividers right? if it is a list view which no divider we don't need to do this calculation and skip?

File: ios/FluentUI/Vnext/List/List.swift [hunk 9]
```diff
@@ -70,12 +70,28 @@ public struct MSFListView: View {
...
+        return lastCell
+    }
+
     private func updateCellDividers() -> [MSFListSectionState] {
         state.sections.forEach { section in
             section.cells.forEach { cell in
                 cell.hasDivider = section.hasDividers
             }
-            section.cells.last?.hasDivider = false
+            if let lastCell = section.cells.last, section.hasDividers {
```
Comment (harrieshin) [inline ios/FluentUI/Vnext/List/List.swift:92]: maybe it is faster a little bit if you switch the order?

**Review by harrieshin — APPROVED** [review]

**Pull request merged**